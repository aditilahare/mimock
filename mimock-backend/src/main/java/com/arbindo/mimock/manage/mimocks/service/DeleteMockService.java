package com.arbindo.mimock.manage.mimocks.service;

public interface DeleteMockService {

    boolean hardDeleteMockById(String mockId);

    boolean softDeleteMockById(String mockId);

    boolean deleteAllMocks();

    void flushDeletedMocks();
}
