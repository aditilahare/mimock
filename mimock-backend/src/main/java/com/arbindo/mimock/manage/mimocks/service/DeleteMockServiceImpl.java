package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.common.services.EntityStatusService;
import com.arbindo.mimock.entities.EntityStatus;
import com.arbindo.mimock.entities.Mock;
import com.arbindo.mimock.repository.MocksRepository;
import com.arbindo.mimock.utils.ValidationUtil;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.time.ZonedDateTime;
import java.util.List;

@Service
@Log4j2
@Builder
@AllArgsConstructor
public class DeleteMockServiceImpl implements DeleteMockService {

    @Autowired
    private MocksRepository mocksRepository;

    @Autowired
    private EntityStatusService entityStatusService;

    @Autowired
    private MockManagementService mockManagementService;

    @Override
    public boolean hardDeleteMockById(String mockId) {
        if (ValidationUtil.isNotNullOrEmpty(mockId)) {
            try {
                Mock mock = mockManagementService.getMockById(mockId);
                if (mock != null) {
                    mocksRepository.delete(mock);
                    return true;
                }
            } catch (Exception e) {
                log.log(Level.DEBUG, e.getMessage());
                return false;
            }
        }
        log.log(Level.DEBUG, "Invalid Mock Id!");
        return false;
    }

    @Transactional
    @Override
    public boolean softDeleteMockById(String mockId) {
        if (ValidationUtil.isNotNullOrEmpty(mockId)) {
            try {
                Mock mock = mockManagementService.getMockById(mockId);
                if (mock != null) {
                    // Perform only soft delete i.e. Mark EntityStatus as DELETED
                    EntityStatus entityStatus = entityStatusService.getDeletedMockEntityStatus();
                    mock.setEntityStatus(entityStatus);
                    mock.setDeletedAt(ZonedDateTime.now());
                    mocksRepository.save(mock);
                    return true;
                }
            } catch (Exception e) {
                log.log(Level.DEBUG, e.getMessage());
                return false;
            }
        }
        log.log(Level.DEBUG, "Invalid Mock Id!");
        return false;
    }

    @Override
    public boolean deleteAllMocks() {
        try {
            mocksRepository.deleteAll();
            return true;
        } catch (Exception e) {
            log.log(Level.DEBUG, e.getMessage());
        }
        log.log(Level.DEBUG, "Unable to delete all mocks!");
        return false;
    }

    @Override
    public void flushDeletedMocks() {
        try {
            EntityStatus entityStatus = entityStatusService.getDeletedMockEntityStatus();
            ZonedDateTime thirtyDaysAgo = ZonedDateTime.now().minusDays(30);
            List<Mock> deletedMocks = mocksRepository.findAllByEntityStatusAndDeletedAt(entityStatus, thirtyDaysAgo);
            mocksRepository.deleteAll(deletedMocks);
            log.log(Level.INFO, "Flushed " + deletedMocks.size() + " mock(s)!");
            return;
        } catch (Exception e) {
            log.log(Level.DEBUG, e.getMessage());
        }
        log.log(Level.DEBUG, "Unable to flush deleted mocks!");
    }
}
