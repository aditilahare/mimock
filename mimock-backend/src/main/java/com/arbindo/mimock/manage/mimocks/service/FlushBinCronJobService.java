package com.arbindo.mimock.manage.mimocks.service;

public interface FlushBinCronJobService {
    void FlushBin();
}
