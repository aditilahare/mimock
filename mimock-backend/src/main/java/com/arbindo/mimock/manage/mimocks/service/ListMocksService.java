package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.entities.Mock;
import com.arbindo.mimock.manage.mimocks.enums.Status;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ListMocksService {

    List<Mock> getAllMocks();

    Page<Mock> getMocksAsPageable(Pageable pageable, Status status, String httpMethod, String expectedResponseType);
}
