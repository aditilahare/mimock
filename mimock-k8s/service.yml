apiVersion: v1
kind: Service
metadata:
  name: mimock-svc
spec:
  type: NodePort
  selector:
    app: mimock
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
      nodePort: 30001
