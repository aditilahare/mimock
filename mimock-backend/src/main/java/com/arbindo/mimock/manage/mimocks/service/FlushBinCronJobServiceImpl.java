package com.arbindo.mimock.manage.mimocks.service;

import com.arbindo.mimock.entities.PlatformSettings;
import com.arbindo.mimock.manage.platformsettings.service.PlatformSettingsService;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@Log4j2
public class FlushBinCronJobServiceImpl implements FlushBinCronJobService {

    @Autowired
    PlatformSettingsService platformSettingsService;

    @Autowired
    DeleteMockService deleteMockService;

    @Override
    @Scheduled(cron = "${flush-bin-cron-expression}")
    public void flushBin() {
        PlatformSettings platformSettings = platformSettingsService.getDefaultPlatformSettings();
        if(!platformSettings.getIsFlushBinCronEnabled()){
            log.log(Level.INFO, "Flush Bin Cron Is Disabled. Skipping CRON scheduler job.");
            return;
        }
        log.log(Level.INFO, "Starting CRON scheduler job.");
        deleteMockService.flushDeletedMocks();
        log.log(Level.INFO, "Completed CRON scheduler job.");
    }
}
