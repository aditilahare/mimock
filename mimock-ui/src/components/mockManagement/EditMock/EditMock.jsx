import React, { useState, useEffect } from 'react';
import { useRecoilState } from 'recoil';
import { useSearchParams } from 'react-router-dom';
import Alert from '@mui/material/Alert';
import AlertTitle from '@mui/material/AlertTitle';
import newMockFieldsAtom from 'atoms/newMockFieldsAtom';
import { getMock } from 'services/mockManagement/getMockById.service';
import { mockManagementConstants } from 'constants/mockManagementConstants';
import AddMock from '../AddMock';

export default function EditMock() {
	const [, setMockData] = useRecoilState(newMockFieldsAtom);
	const [isLoading, setIsLoading] = useState(false);
	const [searchParams] = useSearchParams();
	const [error, setError] = useState({
		isError: false,
		message: '',
	});

	useEffect(() => {
		setIsLoading(true);
		getMock(searchParams.get('mockId'))
			.then((data) => {
				if (
					data.entityStatus !== mockManagementConstants.ENTITY_STATUS.ACTIVE
				) {
					throw new Error('Mock is not ACTIVE');
				}
				setIsLoading(false);
				setMockData({ ...data, mode: mockManagementConstants.mode.EDIT });
			})
			.catch((e) => {
				setIsLoading(false);
				setError({
					isError: true,
					message:
						e.response?.data?.message || e.message || "Couldn't load mock",
				});
			});
	}, []);

	return (
		<If condition={!isLoading}>
			<Choose>
				<When condition={error.isError}>
					<Alert
						data-testid='edit-error-banner'
						severity='error'
						style={{
							marginTop: '2rem',
							width: '60%',
							marginLeft: 'auto',
							marginRight: 'auto',
							padding: '1rem',
						}}
					>
						<AlertTitle>Error</AlertTitle>
						<strong>Failed to edit mock</strong> - {error.message}
					</Alert>
				</When>
				<Otherwise>
					<AddMock mode={mockManagementConstants.mode.EDIT} />
				</Otherwise>
			</Choose>
		</If>
	);
}
