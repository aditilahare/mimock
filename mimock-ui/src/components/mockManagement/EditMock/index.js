import EditMock from './EditMock';

export default EditMock;
