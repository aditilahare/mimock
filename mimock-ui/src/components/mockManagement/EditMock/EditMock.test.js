import React from 'react';
import { act, render } from '@testing-library/react';
import { getMock } from 'services/mockManagement/getMockById.service';
import EditMock from './EditMock';

let mockedRecoilFn;
jest.mock('recoil', () => {
	mockedRecoilFn = jest.fn();
	return {
		atom: jest.fn(),
		useRecoilState: jest.fn(() => [{}, mockedRecoilFn]),
	};
});
jest.mock('../AddMock', () => {
	return function AddMock() {
		return <div data-testid='add-mock'>AddMock</div>;
	};
});
jest.mock('services/mockManagement/getMockById.service');
jest.mock('react-router-dom', () => ({
	useSearchParams: jest.fn(() => [
		{
			get: jest.fn(() => '1'),
		},
	]),
}));

describe('EditMock', () => {
	it('should render edit mock page when mock status is ACTIVE', async () => {
		getMock.mockResolvedValue({
			entityStatus: 'NONE',
		});

		let tree;
		await act(async () => {
			tree = render(<EditMock />);
		});

		const { getByTestId, queryByTestId, container } = tree;

		expect(getMock).toHaveBeenCalledTimes(1);
		expect(getMock).toHaveBeenCalledWith('1');
		expect(getByTestId('add-mock')).toBeInTheDocument();
		expect(queryByTestId('edit-error-banner')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should display error banner when mock status is not ACTIVE', async () => {
		getMock.mockResolvedValue({
			entityStatus: 'ARCHIVED',
		});

		let tree;
		await act(async () => {
			tree = render(<EditMock />);
		});

		const { getByTestId, queryByTestId, getByText, container } = tree;

		expect(getMock).toHaveBeenCalledTimes(1);
		expect(getMock).toHaveBeenCalledWith('1');
		expect(getByTestId('edit-error-banner')).toBeInTheDocument();
		expect(getByText(/Mock is not ACTIVE/)).toBeInTheDocument();
		expect(queryByTestId('add-mock')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should display error banner when fetching mock fails with error', async () => {
		getMock.mockRejectedValue({
			response: {
				data: {
					message: "Mock doesn't exist",
				},
			},
		});

		let tree;
		await act(async () => {
			tree = render(<EditMock />);
		});

		const { getByTestId, getByText, queryByTestId, container } = tree;

		expect(getMock).toHaveBeenCalledTimes(1);
		expect(getMock).toHaveBeenCalledWith('1');
		expect(getByTestId('edit-error-banner')).toBeInTheDocument();
		expect(getByText(/Mock doesn't exist/)).toBeInTheDocument();
		expect(queryByTestId('add-mock')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should display error banner when fetching mock fails with no error data', async () => {
		getMock.mockRejectedValue({});

		let tree;
		await act(async () => {
			tree = render(<EditMock />);
		});

		const { getByTestId, getByText, queryByTestId, container } = tree;

		expect(getMock).toHaveBeenCalledTimes(1);
		expect(getMock).toHaveBeenCalledWith('1');
		expect(getByTestId('edit-error-banner')).toBeInTheDocument();
		expect(getByText(/Couldn't load mock/)).toBeInTheDocument();
		expect(queryByTestId('add-mock')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});
});
