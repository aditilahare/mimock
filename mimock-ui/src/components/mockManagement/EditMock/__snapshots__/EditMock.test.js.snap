// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMock should display error banner when fetching mock fails with error 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
    data-testid="edit-error-banner"
    role="alert"
    style="margin-top: 2rem; width: 60%; margin-left: auto; margin-right: auto; padding: 1rem;"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-acap47-MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root"
      >
        Error
      </div>
      <strong>
        Failed to edit mock
      </strong>
       - 
      Mock doesn't exist
    </div>
  </div>
</div>
`;

exports[`EditMock should display error banner when fetching mock fails with no error data 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
    data-testid="edit-error-banner"
    role="alert"
    style="margin-top: 2rem; width: 60%; margin-left: auto; margin-right: auto; padding: 1rem;"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-acap47-MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root"
      >
        Error
      </div>
      <strong>
        Failed to edit mock
      </strong>
       - 
      Couldn't load mock
    </div>
  </div>
</div>
`;

exports[`EditMock should display error banner when mock status is not ACTIVE 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
    data-testid="edit-error-banner"
    role="alert"
    style="margin-top: 2rem; width: 60%; margin-left: auto; margin-right: auto; padding: 1rem;"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-acap47-MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root"
      >
        Error
      </div>
      <strong>
        Failed to edit mock
      </strong>
       - 
      Mock is not ACTIVE
    </div>
  </div>
</div>
`;

exports[`EditMock should render edit mock page when mock status is ACTIVE 1`] = `
<div>
  <div
    data-testid="add-mock"
  >
    AddMock
  </div>
</div>
`;
